addons/binding/org.openhab.binding.verisure/README.md

```diff
@@ -0,0 +1,285 @@
+# Verisure Binding
+
+This is an OpenHAB binding for Versiure Alarm system, by Securitas Direct.
+
+This binding uses the rest API behind the myverisure pages https://mypages.verisure.com/login.html.
+
+The binding supports several installation sites via the configuration parameter numberOfInstallations that defaults to 1.
+
+Be aware that Verisure don't approve if you update to often, I have gotten no complaints running with a 10 minutes update interval, but officially you should use 30 minutes.
+
+## Supported Things
+
+This binding supports the following thing types:
+
+- Bridge
+- Alarm
+- Smoke Detector (climate) 
+- Water Detector (climate)
+- Siren (climate)
+- Night Control
+- Yaleman SmartLock
+- SmartPlug
+- Door/Window Status
+- User Presence Status
+- Broadband Connection Status
+
+
+## Binding Configuration
+
+You will have to configure the bridge with username and password, these must be the same values as used when logging into https://mypages.verisure.com. 
+You can also configure your pin-code to be able to lock/unlock the SmartLock and arm/unarm the alarm. 
+It is also possible to configure the number of Verisure installations/sites you have, default is 1.
+
+## Discovery
+
+After the configuration of the Verisure Bridge all of the available Sensors, Alarms, SmartPlugs, SmartLocks and Climate devices will be discovered and placed as things in the inbox.
+
+## Thing Configuration
+
+Only the bridge require manual configuration. The devices and sensors should not be added by hand, let the discovery/inbox initially configure these.
```

Comment:
Manual adding should be supported as well. Currently thing id is a Verisure device id. Verisure device id could/should be thing parameter, then user could give better name for the thing (e.g. frontDoorLock).

Reply:
It is supported to add things manually, I will refrase the wording to:
"The devices and sensors can be added by hand, or you can let the discovery mechanism automatically find all of your Verisure things."
OK?

I think it is relevant to use the Versure ID as the thing ID, the user is familiar with it since it is used when physically installing the Verisure device.

The user will also configure a thing/item name that then will be used in the openHAB installation.

Reply:
I have 3 Verisure locks and I don't remember my locks id's :)

Most of the bindings have also some kind of device id/address which need to be linked somehow to Thing. 

Just one example:

```
Bridge rfxcom:tcpbridge:garage [ host="sunflower", port=10001 ] {
    Thing temperaturehumidity OutdoorSensor [deviceId="100001.1", subType="AC"]
}
``` 
So Thing name 
`rfxcom:temperaturehumidity:garage:OutdoorSensor `
compared to
`rfxcom:temperaturehumidity:123456:346564 `
is much more user friendly, at least when you have more than one device. If user want to use id's, it is still possible.

Channel parameters is really simple concept to implement. At least JSON for locks from Verisure contains name of the device/location which user have defined during installation. So autodiscovery could name thing by location + id and then id is but as Thing parameter. Thing handler will use channel parameter rather than thing name to correlate the data to thing from JSON. Then Things can be added manually, so that name of the thing is "freely" selectable by the user. 

WDYT?

Reply:
> I have 3 Verisure locks and I don't remember my locks id's :)

Cool :-)

> So Thing name
> `rfxcom:temperaturehumidity:garage:OutdoorSensor `
> compared to
> `rfxcom:temperaturehumidity:123456:346564 `
> is much more user friendly, at least when you have more than one device. If user want to use id's, it is still possible.

I get your point and I agree that it could be easier when configuring Things manually.

> Channel parameters is really simple concept to implement. 

You then mention both "Channel parameters" and "Thing parameter", could you give an example of configuration of such a parameter in a things-types.xml? Preferably as a part of my things-types.xml file.

> So autodiscovery could name thing by location + id and then id is but as Thing parameter. 

Could you elaborate on this, I dont understand the whole sentence?  When you write "name", is it the Label you mean then? When you write "id", what id are you then referring to? 
"Thing parameter" here, should that be channel parameter? 
Today the implementation sets the Label during autodiscovery to Verisure ID and Location if present and ThingUID is constructed using Thing id set to Verisure ID. 

>Thing handler will use channel parameter rather than thing name to correlate the data to thing from JSON. Then Things can be added manually, so that name of the thing is "freely" selectable by the user.

Same here, what do you mean by "thing name"?

So what I think you mean is:

-  If autodiscovery is used, Thing id should be configured as today
-  If manual configuration is used, Thing id should be set to anything the user wirites
- ThingHandler should not use Thing Id for correlation, instead use a new channel parameter that is set to Verisure ID (as Thing id is set today)

If that is correct, in what call scenario should the new channel parameter be set to Verisure ID?

Reply:
> So what I think you mean is:
> 
> If autodiscovery is used, Thing id should be configured as today
> If manual configuration is used, Thing id should be set to anything the user wirites
> ThingHandler should not use Thing Id for correlation, instead use a new channel parameter that is set to Verisure ID (as Thing id is set today)
> If that is correct, in what call scenario should the new channel parameter be set to Verisure ID?

When autodiscovery is used, things can be generated as today, but discovery should add device id (verisure id) to thing parameters. Then Thing handler doesn't need to have different implementation for discovered and manually added devices. So, thing handler should not be interest-in the thing name, but should always take device/verisure id from thing parameters.

As I already referent to rfxcom binding and temperature humidity sensor, let's use that as an example.

Temperaturehumidity thing definition:
https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.rfxcom/ESH-INF/thing/temperaturehumidity.xml
Starting from line 26, you can find config-description which contains deviceId (similar than verisure id).

Then you can 
1. introduce java class which present the thing parameters
2. Or ask thing parameters directly by parameter name.

Rfxcom binding use method 1:
See https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/config/RFXComDeviceConfiguration.java

In ThingHandler you just ask thing configuration and use those parameter. See lines 104 and 105 from https://github.com/openhab/openhab2-addons/blob/master/addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/handler/RFXComHandler.java.

Hopefully this make sense and don't hesitate to ask if there is still something unclear.

Reply:
So then a "Thing parameter" in your binding is a mandatory configuration parameter for a Thing that is set during autodiscovery when that is being used. 

But when you use manual configuration you will still need to know the Verisure ID since that needs to be configured as a Thing property:

`Thing smartLock SmartLockEntre "Entré Lås"  [ ID="2A8L_XXXX" ] `

So during autodiscovery the "Thing name" or Thing UID will be `verisure:smartLock:1:2A8L_XXXX` and when using manual configuration it will be `verisure:smartLock:1:SmartLockEntre`.

Autodiscovery:
- Thing ID = Verisure ID

Manual config:
- Thing ID = Manual configured

Then use new Thing configuration property for for correlation in ThingHandlers instead of Thing ID.

Correct?

So if this is correct you still need to know the Verisure ID when configuring manually, the Thing UID can then contain a manually configured Thing ID. 

When creating items in an .items file you need to know the channel ID for the Thing's channel, a part of that is the Thing UID. However, I always use the copy function from Paper UI so I do not really see the big gain in having Thing ID set to `smartLockEntre`compared to `2A8L_XXXX`, or am I missing something?

Reply:
Yes, I believe you understand what I meant. But, what I said in my first comment, autodiscovery could also make better name during discovery for thing, because discovery knows area, type, etc.

Autodiscovery:
Thing ID = Type + Area + Verisure ID -> Lock-Frontdoor-12345

ThingUID need to be unique, so auto discovery could still use just verisure id like you have used already.

Similar concept is used pretty much in every binding. See e.g. [Developing a Binding for openHAB](https://www.openhab.org/docs/developer/development/bindings.html). 

Where location is like a device id
`yahooweather:weather:berlin     [ location="638242" ]`

You need to know your location when setup things, but you still can give better name for thing.

As a user, I like to configure my things by configuration files, so I don't use e.g. Paper UI. If I add my locks manually, I need to know that ThingUID in this binding (different for other bindings) need to be "exactly" the verisure id. Well, spaces need to be replaced by underscore, etc. How about letters, should they be in upper case or is lower case also supported. So start be pretty cumbersome, right.

Reply:
> ThingUID need to be unique, so auto discovery could still use just verisure id like you have used already.

Yes, I agree that it is best to keep Verisure ID as Thing ID for autodiscovery, but the Label (Name) could be set maybe as today with additional information about Locaction for things that has that parameter.

> Similar concept is used pretty much in every binding. See e.g. [Developing a Binding for openHAB](https://www.openhab.org/docs/developer/development/bindings.html).
> 
> Where location is like a device id
> `yahooweather:weather:berlin [ location="638242" ]`
> 
> You need to know your location when setup things, but you still can give better name for thing.

I get the point, and I'm still in a learning phase :-) and since it is the standard procedure, the Verisure binding shall be adapted to follow it.

> As a user, I like to configure my things by configuration files, so I don't use e.g. Paper UI. If I add my locks manually, I need to know that ThingUID in this binding (different for other bindings) need to be "exactly" the verisure id. Well, spaces need to be replaced by underscore, etc. How about letters, should they be in upper case or is lower case also supported. So start be pretty cumbersome, right.

I guess I have too little experience of defining Things in configuration files, I use a lot of configuration files for items and site maps, but I really like the autodiscovery feature since I think that is a big selling point for openHAB and it is easier to use for newbies as well.

I will now try to adapt the Verisure binding as you have described, thanks for your help and explanation and for pointing me in the correct direction! :-) 

I'll post here when I have pushed the updated and adapted code.

Reply:
I've now pushed a commit with the changes that @paulianttila suggested to adapt the implementation to the Thing Id naming standard, the README is updated with the new configuration property `deviceId`and examples have been added with manual configuration using user defined Thing Id's. Verisure Id is used to populate deviceId except for Alarm, Broadband Connection and User Presence that lacks a Verisure Id.

I've updated the [community post](https://community.openhab.org/t/testers-for-verisure-openhab-2-binding/7499/134?u=jannegpriv) with a new jar-file and links to updated README.

@Hilbrand I tried to squash this last commit with the initial commit but I ended up in for me unresolveable git conflicts during the rebase operation, I would have loved to be able to fix them but I really need some help in doing the squash operation if needed. I'll add you as a collaborator to my branch.

Reply:
I quickly tested the latest version (this is my first time) and at least my smart lock things went immediately online :+1:

About the squash, you don't need to squash your commits. When PR will be merged to master, all you commits will be committed to master by one. From review point of view, it's actually easier to see modifications if fixes are in separated commits.

Reply:
Cool :-)

Reply:
@jannegpriv What @paulianttila says. You don't need to squash. However if you would, I can still clean up your branch and remove the merge commits, but keep all your own commits.

Reply:
OK, I'm happy without squashing my commits :-)
I need to improve my git rebase knowledge!